cli/commands/scaffold/action.go

```diff
@@ -0,0 +1,351 @@
+package scaffold
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/gruntwork-io/terragrunt/config"
+	"github.com/gruntwork-io/terragrunt/shell"
+
+	"github.com/gruntwork-io/terragrunt/terraform"
+
+	"github.com/gruntwork-io/terragrunt/cli/commands/hclfmt"
+	"github.com/gruntwork-io/terragrunt/util"
+
+	boilerplate_options "github.com/gruntwork-io/boilerplate/options"
+	"github.com/gruntwork-io/boilerplate/templates"
+	"github.com/gruntwork-io/boilerplate/variables"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terratest/modules/files"
+	"github.com/hashicorp/go-getter"
+)
+
+const (
+	moduleCount            = 2
+	moduleAndTemplateCount = 3
+
+	sourceUrlTypeHttps = "git-https"
+	sourceUrlTypeGit   = "git-ssh"
+	sourceGitSshUser   = "git"
+
+	sourceUrlTypeVar    = "SourceUrlType"
+	sourceGitSshUserVar = "SourceGitSshUser"
+	refVar              = "Ref"
+	// refParam - ?ref param from url
+	refParam = "ref"
+
+	moduleUrlPattern = `git::([^:]+)://([^/]+)(/.*)`
+	moduleUrlParts   = 4
+
+	defaultBoilerplateConfig = `
+variables:
+
+`
+	defaultTerragruntTemplate = `
+# This is a Terragrunt module generated by boilerplate.
+terraform {
+  source = "{{ .sourceUrl }}"
+}
+
+inputs = {
+  # --------------------------------------------------------------------------------------------------------------------
+  # Required input variables
+  # --------------------------------------------------------------------------------------------------------------------
+  {{ range .requiredVariables }}
+  # Description: {{ .Description }}
+  # Type: {{ .Type }}
+  {{ .Name }} = {{ .DefaultValuePlaceholder }}  # TODO: fill in value
+  {{ end }}
+
+  # --------------------------------------------------------------------------------------------------------------------
+  # Optional input variables
+  # Uncomment the ones you wish to set
+  # --------------------------------------------------------------------------------------------------------------------
+  {{ range .optionalVariables }}
+  # Description: {{ .Description }}
+  # Type: {{ .Type }}
+  # {{ .Name }} = {{ .DefaultValue }}
+  {{ end }}
+}
+`
+)
+
+var moduleUrlRegex = regexp.MustCompile(moduleUrlPattern)
+
+func Run(opts *options.TerragruntOptions) error {
+	// download remote repo to local
+	var moduleUrl = ""
+	var templateUrl = ""
+	// clean all temp dirs
+	var dirsToClean []string
+	defer func() {
+		for _, dir := range dirsToClean {
+			if err := os.RemoveAll(dir); err != nil {
+				opts.Logger.Warnf("Failed to clean up dir %s: %v", dir, err)
+			}
+		}
+	}()
+
+	// scaffold only in empty directories
```

Comment:
Why only in empty dirs? I think scaffolding into an existing dir is OK?

I suppose we could have a warning / confirmation prompt (unless `--non-interactive` is set) telling the user the existing contents may be overwritten, but even that is a bit of a nice to have...

Reply:
Added warning message